Конструктор компаний

Wellheor1 opened the pull request:
## Описание изменений
* Добавление/редактирование компаний

Wellheor1 committed:
Create

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
Загрузка списка компаний + Разметка

* Функция запроса списка компаний
* flex
* Разметка

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
Договоры, редактирование, поиск

* Редактирование компаний
* Очистка редактора компаний
* Поиск по компаниям
* Список договоров

Wellheor1 committed:
Конструктор компаний v1

* Создание/редактирование компаний
* Логи

Wellheor1 committed:
black fix

Wellheor1 committed:
Данные компании + уникальность

* Получение данных конкретной компании
* Проверка на уникальность названия компании

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
black fix

Wellheor1 committed:
fix name

Wellheor1 committed:
.

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,296 @@
+<template>
```
Давай форму на Formulate переделаем
Тут оно лучше подойдет
Пример в ConstructOrg

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,296 @@
+          maxlength="12"
+          placeholder="Введите ИНН"
+          class="form-control"
+          @input="getOnlyNumber($event, 'editCompanyInn')"
```
Это можно было решить через модификатор .number на v-model
С точки зрения использования Formulate будет не так актуально, там https://vueformulate.com/guide/inputs/types/text/#number
Там будет важным пункт про inputmode="numeric", тип поля лучше оставить текстовый, что бы не было стрелочек + и -

mikhailprivalov reviewed (commented):
Нужно ещё добавить валидацию клиентскую и серверную, что нужные поля заполнены.
Ну точнее валидация на клиенте есть, но через formulare ее можно легко сделать realtime.
Плюс все валидации на клиенте надо дублировать на бекенде.

mikhailprivalov reviewed (commented):
Так же стоит добавить validation="required:trim" на поля FormulateInput

mikhailprivalov commented on api/views.py:
```diff
@@ -2512,3 +2513,129 @@ def delete_research_in_price(request):
+        }
+        for company in Company.objects.filter(active_status=True)
+    ]
+    company_data = sorted(company_data, key=lambda company: company["title"])
```
Можно было через order_by в запросе

mikhailprivalov commented on api/views.py:
```diff
@@ -2512,3 +2513,129 @@ def delete_research_in_price(request):
+        }
+        for contract in Contract.objects.filter(active_status=True)
+    ]
+    contract_data = sorted(contract_data, key=lambda contract: contract["label"])
```
Тут тоже order_by

mikhailprivalov commented on api/views.py:
```diff
@@ -2512,3 +2513,129 @@ def delete_research_in_price(request):
+        "ogrn": current_company.ogrn,
+        "kpp": current_company.kpp,
+        "bik": current_company.bik,
+        "contract_id": current_company.contract.pk,
```
Лучше будет
```suggestion
        "contract_id": current_company.contract_id,
```

Иначе:
1. Это дополнительный запрос в базу
2. Если contract пуст, то должна быть ошибка

mikhailprivalov replied:
2 - важно, потому что поле contract в компании null=True

mikhailprivalov commented on api/views.py:
```diff
@@ -2512,3 +2513,129 @@ def delete_research_in_price(request):
+            company_data.pk,
+            130002,
+            request.user.doctorprofile,
+            {"old_company_data": old_company_data, "new_company_data": new_company_data},
```
для old_company_data и new_company_data можно добавить функцию as_json в модель Company, что бы четыре раза не делать одно и то же (по всему новому коду)

mikhailprivalov commented on api/views.py:
```diff
@@ -2512,3 +2513,129 @@ def delete_research_in_price(request):
+        }
+        Log.log(
+            company_data.pk,
+            130002,
```
Так как body в логах на создание и редактирование разные - лучше сделать разные события добавления и обновления

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,296 @@
+    },
+    async editCompany(company) {
+      await this.$store.dispatch(actions.INC_LOADING);
+      this.currentCompany = await this.$api('get-current-company', { id: company.id });
```
Можно так
```suggestion
      this.currentCompany = await this.$api('get-current-company', company, 'id');
```

mikhailprivalov replied:
api имеет такие варианты применения:
this.$api('get-current-company', { id: company.id })
this.$api('get-current-company', company, 'id');
this.$api('get-current-company', company, ['id', 'id2']);
this.$api('get-current-company', company, 'id', {moreData: 1});
this.$api('get-current-company', company, ['id', 'id2'], {moreData: 1});

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,296 @@
+        Компании
+      </h5>
+      <input
+        v-model="search"
```
Такие поля лучше очищать от ненужных пробелов
```suggestion
        v-model.trim="search"
```

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,296 @@
+        this.$root.$emit('msg', 'error', 'Не заполнено название');
+      } else if (this.editCompanyContractId === -1) {
+        this.$root.$emit('msg', 'error', 'Не выбран договор');
+      } else if (this.editCompanyId === -1 && this.dataCompanyList.find((company) => company.title === this.editCompanyTitle)) {
```
Условие не полноценное.
Если мы меняем компанию и делаем название уже существующее, то будут дубликаты.
Плюс дубликацию проверять можно ещё по ИНН

mikhailprivalov reviewed (changes requested):
Есть вопросы

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
Update api/views.py

Co-authored-by: Mikhail Privalov <n5250@ya.ru>

Wellheor1 committed:
Update l2-frontend/src/construct/ConstructCompany.vue

Co-authored-by: Mikhail Privalov <n5250@ya.ru>

Wellheor1 committed:
fix

Wellheor1 committed:
Merge branch 'Company' of https://github.com/moodpulse/l2 into Company

Wellheor1 committed:
fix

mikhailprivalov commented on api/urls.py:
```diff
@@ -101,4 +101,8 @@
     path('delete-research-in-price', views.delete_research_in_price),
+    path('get-company-list', views.get_company_list),
+    path('get-contract-list', views.get_contract_list),
+    path('get-current-company', views.get_current_company),
```
Почему get current, а не просто get?

Wellheor1 committed:
.

Wellheor1 committed:
.

Wellheor1 committed:
.

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
.

Wellheor1 committed:
Конструктор компаний v2

* Редактирование - через Formulate
* Валидация

Wellheor1 committed:
Update contracts/models.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Wellheor1 committed:
Update api/views.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Wellheor1 committed:
fix style

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
.

Wellheor1 committed:
.

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
.

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
.

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
.

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+      </h5>
+      <input
+        v-model.trim="search"
+        style="margin-top: 36px; padding-left: 0.75rem;"
```
Почему паддинги в rem?

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+            pattern="[0-9]*"
+          />
+          <label>Договор</label>
+          <Treeselect
```
А если всё же formulate select?

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+      this.currentCompany = await this.$api('get-company', company, 'pk');
+      await this.$store.dispatch(actions.DEC_LOADING);
+      this.dataCurrentCompany = this.currentCompany.data;
+      this.editorCompany = this.dataCurrentCompany;
```
Что тут происходит?) 
Зачем столько перекладываний туда сюда и полей в data?)

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  padding-left: 0.75rem;
```
Не надо rem
Используй пиксели всегда, когда это возможно.

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+  display: flex;
+  justify-content: flex-end;
+}
+label {
```
Что за стиль и для чего? 
Не стоит такие общие селекторы писать
Нужно давать классы

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+    </div>
+    <div class="box card-1 card-no-hover">
+      <h5 class="text-center">
+        {{ editorCompany.pk ? 'Обновить компанию' : 'Добавить компанию' }}
```
Добавь такое computed свойство
isNewCompany, которое внутри себя будет проверять на pk.
В разметке код будет лучше.

mikhailprivalov reviewed (changes requested):
Есть вопросы. И прикрепи плиз скриншот результата

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+      <h5 class="text-center">
+        {{ editorCompany.pk ? 'Обновить компанию' : 'Добавить компанию' }}
+      </h5>
+      <h6 class="text-center">
```
Почему это всегда отображаем?

mikhailprivalov replied:
Надо на обновлении только

mikhailprivalov commented on l2-frontend/src/construct/ConstructCompany.vue:
```diff
@@ -0,0 +1,285 @@
+        if (ok) {
+          this.$root.$emit('msg', 'ok', 'Сохранено');
+          await this.getCompanyList();
+          this.clearEditCompany();
```
Думаю это лучше вызвать до получения списка. У тебя там await, выходит форма очистится с лагом

Wellheor1 committed:
fix

urchinpro commented on api/views.py:
```diff
@@ -2517,3 +2518,93 @@ def delete_research_in_price(request):
+        company_data.legal_address = request_data["legalAddress"]
+        company_data.fact_address = request_data["factAddress"]
+        company_data.inn = request_data["inn"]
+        company_data.ogrn = request_data["ogrn"]
```
orgn - заменить на ogrn

urchinpro replied:
везде

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 committed:
fix

Wellheor1 committed:
contractData

Вывод только не используемых договоров
Добавление текущего договора компании во время редактирования

Wellheor1 committed:
Merge branch 'develop' into Company

Wellheor1 commented:
![image](https://user-images.githubusercontent.com/72241044/199363066-864d6863-37fd-4d5c-9422-dc66949c0f2f.png)
![image](https://user-images.githubusercontent.com/72241044/199363102-e9ea8adf-c9bb-475c-835e-365a321a6121.png)

Wellheor1 committed:
.

Wellheor1 committed:
Merge branch 'Company' of https://github.com/moodpulse/l2 into Company
